instrumentation/opentelemetry-instrumentation-openai/src/opentelemetry/instrumentation/openai/patch.py

```diff
@@ -0,0 +1,438 @@
+        if hasattr(item, "tool_calls") and item.tool_calls is not None:
+            return item.tool_calls
+        return None
+
+
+@silently_fail
+def _set_input_attributes(span, kwargs, attributes):
+    tools = []
+    for field, value in attributes.model_dump(by_alias=True).items():
+        set_span_attribute(span, field, value)
+
+    if kwargs.get("functions") is not None and kwargs.get("functions") != NOT_GIVEN:
+        for function in kwargs.get("functions"):
+            tools.append(json.dumps({"type": "function", "function": function}))
+
+    if kwargs.get("tools") is not None and kwargs.get("tools") != NOT_GIVEN:
+        tools.append(json.dumps(kwargs.get("tools")))
+
+    if tools:
+        set_span_attribute(span, SpanAttributes.LLM_TOOLS, json.dumps(tools))
+
+
+@silently_fail
+def _set_response_attributes(span, kwargs, result):
+    set_span_attribute(span, SpanAttributes.LLM_RESPONSE_MODEL, result.model)
+    if hasattr(result, "choices") and result.choices is not None:
+        responses = [
+            {
+                "role": (
+                    choice.message.role
+                    if choice.message and choice.message.role
+                    else "assistant"
+                ),
+                "content": extract_content(choice),
+                **(
+                    {"content_filter_results": choice["content_filter_results"]}
+                    if "content_filter_results" in choice
+                    else {}
+                ),
+            }
+            for choice in result.choices
+        ]
+        set_event_completion(span, responses)
+
+    if (
+        hasattr(result, "system_fingerprint")
+        and result.system_fingerprint is not None
+        and result.system_fingerprint != NOT_GIVEN
+    ):
+        set_span_attribute(
+            span,
+            SpanAttributes.LLM_SYSTEM_FINGERPRINT,
+            result.system_fingerprint,
+        )
+    # Get the usage
+    if hasattr(result, "usage") and result.usage is not None:
+        usage = result.usage
+        if usage is not None:
+            set_span_attribute(
+                span,
+                SpanAttributes.LLM_USAGE_PROMPT_TOKENS,
+                result.usage.prompt_tokens,
+            )
+            set_span_attribute(
+                span,
+                SpanAttributes.LLM_USAGE_COMPLETION_TOKENS,
+                result.usage.completion_tokens,
+            )
+            set_span_attribute(
+                span,
+                SpanAttributes.LLM_USAGE_TOTAL_TOKENS,
+                result.usage.total_tokens,
+            )
+
+
+def set_event_prompt(span: Span, prompt):
+    span.add_event(
+        name=SpanAttributes.LLM_CONTENT_PROMPT,
+        attributes={
+            SpanAttributes.LLM_PROMPTS: prompt,
+        },
+    )
+
+
+def set_span_attributes(span: Span, attributes: dict):
+    for field, value in attributes.model_dump(by_alias=True).items():
+        set_span_attribute(span, field, value)
```

Comment:
which attributes it will produce? what would be their names?

we should record attributes documented in the spec https://github.com/open-telemetry/semantic-conventions/blob/v1.27.0/docs/gen-ai/gen-ai-spans.md#genai-attributes

Reply:
@lmolkova These are coming from LLMSpanAttributes feed with SpanAttributes from https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2759/files#diff-0042eaf389fe22a3c3045eda672af6ece462798488c3c3712e99b0727a114f45R21